## Import libraries

In [2]:
import pandas as pd
import numpy as np

## Read the data

In [5]:
df = pd.read_csv("raw_data/nls97.csv")

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8984 entries, 0 to 8983
Data columns (total 89 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   personid               8984 non-null   int64  
 1   gender                 8984 non-null   object 
 2   birthmonth             8984 non-null   int64  
 3   birthyear              8984 non-null   int64  
 4   highestgradecompleted  6663 non-null   float64
 5   maritalstatus          6672 non-null   object 
 6   childathome            4791 non-null   float64
 7   childnotathome         4791 non-null   float64
 8   wageincome             5091 non-null   float64
 9   weeklyhrscomputer      5792 non-null   object 
 10  weeklyhrstv            6711 non-null   object 
 11  nightlyhrssleep        6706 non-null   float64
 12  satverbal              1406 non-null   float64
 13  satmath                1407 non-null   float64
 14  gpaoverall             6004 non-null   float64
 15  gpae

## Set the "personid" column as the index

In [10]:
df.set_index("personid", inplace = True)

In [12]:
df.head(3)

gender  birthmonth  birthyear  highestgradecompleted  maritalstatus  \
personid                                                                        
100061    Female           5       1980                   13.0        Married   
100139      Male           9       1983                   12.0        Married   
100284      Male          11       1984                    7.0  Never-married   

          childathome  childnotathome  wageincome        weeklyhrscomputer  \
personid                                                                     
100061            4.0             0.0     12500.0  10 hours or more a week   
100139            2.0             0.0    120000.0      1 to 3 hours a week   
100284            1.0             0.0     58000.0                      NaN   

                    weeklyhrstv  ...      colenrfeb13      colenroct13  \
personid                         ...                                     
100061    11 to 20 hours a week  ...  1. Not enrolled  1. Not enrolled   
100139     3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   
100284    11 to 20 hours a week  ...  1. Not enrolled  1. Not enrolled   

              colenrfeb14      colenroct14      colenrfeb15      colenroct15  \
personid                                                                       
100061    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
100139    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
100284    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   

              colenrfeb16      colenroct16      colenrfeb17      colenroct17  
personid                                                                      
100061    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  
100139    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  
100284    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  

[3 rows x 88 columns]

## Selecting the rows by Slicing the fifth to eighth rows

In [15]:
df[4:8] #fifth row is index 4

gender  birthmonth  birthyear  highestgradecompleted  maritalstatus  \
personid                                                                        
100583      Male           1       1980                   13.0        Married   
100833      Male           1       1981                   12.0        Married   
100931    Female           8       1983                   15.0       Divorced   
101089      Male           4       1980                   13.0  Never-married   

          childathome  childnotathome  wageincome        weeklyhrscomputer  \
personid                                                                     
100583            4.0             0.0     30000.0  Less than 1 hour a week   
100833            2.0             0.0     39000.0      1 to 3 hours a week   
100931            2.0             0.0     56000.0  10 hours or more a week   
101089            NaN             NaN     36000.0      4 to 6 hours a week   

                         weeklyhrstv  ...         colenrfeb13  \
personid                              ...                       
100583          3 to 10 hours a week  ...     1. Not enrolled   
100833          3 to 10 hours a week  ...     1. Not enrolled   
100931    Less than 2 hours per week  ...  2. 2-year college    
101089    Less than 2 hours per week  ...     1. Not enrolled   

                 colenroct13         colenrfeb14         colenroct14  \
personid                                                               
100583       1. Not enrolled     1. Not enrolled     1. Not enrolled   
100833       1. Not enrolled     1. Not enrolled     1. Not enrolled   
100931    2. 2-year college      1. Not enrolled     1. Not enrolled   
101089       1. Not enrolled  2. 2-year college   2. 2-year college    

                 colenrfeb15      colenroct15      colenrfeb16  \
personid                                                         
100583       1. Not enrolled  1. Not enrolled  1. Not enrolled   
100833       1. Not enrolled  1. Not enrolled  1. Not enrolled   
100931       1. Not enrolled  1. Not enrolled  1. Not enrolled   
101089    2. 2-year college   1. Not enrolled  1. Not enrolled   

              colenroct16      colenrfeb17      colenroct17  
personid                                                     
100583    1. Not enrolled  1. Not enrolled  1. Not enrolled  
100833    1. Not enrolled  1. Not enrolled  1. Not enrolled  
100931    1. Not enrolled  1. Not enrolled  1. Not enrolled  
101089    1. Not enrolled  1. Not enrolled  1. Not enrolled  

[4 rows x 88 columns]

## Transpose of the sliced data

In [18]:
df[4:8].T

personid                        100583           100833           100931  \
gender                            Male             Male           Female   
birthmonth                           1                1                8   
birthyear                         1980             1981             1983   
highestgradecompleted             13.0             12.0             15.0   
maritalstatus                  Married          Married         Divorced   
...                                ...              ...              ...   
colenroct15            1. Not enrolled  1. Not enrolled  1. Not enrolled   
colenrfeb16            1. Not enrolled  1. Not enrolled  1. Not enrolled   
colenroct16            1. Not enrolled  1. Not enrolled  1. Not enrolled   
colenrfeb17            1. Not enrolled  1. Not enrolled  1. Not enrolled   
colenroct17            1. Not enrolled  1. Not enrolled  1. Not enrolled   

personid                        101089  
gender                            Male  
birthmonth                           4  
birthyear                         1980  
highestgradecompleted             13.0  
maritalstatus            Never-married  
...                                ...  
colenroct15            1. Not enrolled  
colenrfeb16            1. Not enrolled  
colenroct16            1. Not enrolled  
colenrfeb17            1. Not enrolled  
colenroct17            1. Not enrolled  

[88 rows x 4 columns]

## Slicing with a step

In [21]:
df[4:14:2] # Selecting every other row starting from index 4 (row 5)

gender  birthmonth  birthyear  highestgradecompleted  maritalstatus  \
personid                                                                        
100583      Male           1       1980                   13.0        Married   
100931    Female           8       1983                   15.0       Divorced   
101122      Male           2       1981                    NaN            NaN   
101230    Female           2       1984                   11.0  Never-married   
101414      Male           6       1984                   12.0  Never-married   

          childathome  childnotathome  wageincome        weeklyhrscomputer  \
personid                                                                     
100583            4.0             0.0     30000.0  Less than 1 hour a week   
100931            2.0             0.0     56000.0  10 hours or more a week   
101122            NaN             NaN         NaN                      NaN   
101230            2.0             0.0         NaN                      NaN   
101414            2.0             0.0     20000.0                      NaN   

                         weeklyhrstv  ...         colenrfeb13  \
personid                              ...                       
100583          3 to 10 hours a week  ...     1. Not enrolled   
100931    Less than 2 hours per week  ...  2. 2-year college    
101122                           NaN  ...     1. Not enrolled   
101230    Less than 2 hours per week  ...     1. Not enrolled   
101414          3 to 10 hours a week  ...     1. Not enrolled   

                 colenroct13      colenrfeb14      colenroct14  \
personid                                                         
100583       1. Not enrolled  1. Not enrolled  1. Not enrolled   
100931    2. 2-year college   1. Not enrolled  1. Not enrolled   
101122       1. Not enrolled  1. Not enrolled  1. Not enrolled   
101230       1. Not enrolled  1. Not enrolled  1. Not enrolled   
101414       1. Not enrolled  1. Not enrolled  1. Not enrolled   

              colenrfeb15      colenroct15      colenrfeb16      colenroct16  \
personid                                                                       
100583    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
100931    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
101122    1. Not enrolled  1. Not enrolled              NaN              NaN   
101230    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
101414    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   

              colenrfeb17      colenroct17  
personid                                    
100583    1. Not enrolled  1. Not enrolled  
100931    1. Not enrolled  1. Not enrolled  
101122                NaN              NaN  
101230    1. Not enrolled  1. Not enrolled  
101414    1. Not enrolled  1. Not enrolled  

[5 rows x 88 columns]

## Slicing the last three rows of the dataframe using "tail"

In [24]:
df.tail(3)

gender  birthmonth  birthyear  highestgradecompleted  maritalstatus  \
personid                                                                        
999543    Female           8       1984                   12.0       Divorced   
999698    Female           5       1983                   12.0  Never-married   
999963    Female           9       1982                   17.0        Married   

          childathome  childnotathome  wageincome    weeklyhrscomputer  \
personid                                                                 
999543            2.0             0.0         NaN                  NaN   
999698            5.0             0.0         NaN  4 to 6 hours a week   
999963            2.0             0.0     50000.0  4 to 6 hours a week   

                        weeklyhrstv  ...      colenrfeb13      colenroct13  \
personid                             ...                                     
999543         3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   
999698    More than 40 hours a week  ...  1. Not enrolled  1. Not enrolled   
999963         3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   

              colenrfeb14      colenroct14      colenrfeb15      colenroct15  \
personid                                                                       
999543    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
999698    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
999963    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   

              colenrfeb16      colenroct16      colenrfeb17      colenroct17  
personid                                                                      
999543    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  
999698    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  
999963    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  

[3 rows x 88 columns]

## Slicing the last three rows of the dataframe using "negative index"

In [27]:
df[-3:]

gender  birthmonth  birthyear  highestgradecompleted  maritalstatus  \
personid                                                                        
999543    Female           8       1984                   12.0       Divorced   
999698    Female           5       1983                   12.0  Never-married   
999963    Female           9       1982                   17.0        Married   

          childathome  childnotathome  wageincome    weeklyhrscomputer  \
personid                                                                 
999543            2.0             0.0         NaN                  NaN   
999698            5.0             0.0         NaN  4 to 6 hours a week   
999963            2.0             0.0     50000.0  4 to 6 hours a week   

                        weeklyhrstv  ...      colenrfeb13      colenroct13  \
personid                             ...                                     
999543         3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   
999698    More than 40 hours a week  ...  1. Not enrolled  1. Not enrolled   
999963         3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   

              colenrfeb14      colenroct14      colenrfeb15      colenroct15  \
personid                                                                       
999543    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
999698    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
999963    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   

              colenrfeb16      colenroct16      colenrfeb17      colenroct17  
personid                                                                      
999543    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  
999698    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  
999963    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  

[3 rows x 88 columns]

## Using "loc" to select few rows by index label

In [32]:
df.loc[195884:195970] #passing a range of labels

gender  birthmonth  birthyear  highestgradecompleted  maritalstatus  \
personid                                                                        
195884      Male          12       1981                    NaN            NaN   
195891      Male           9       1980                   12.0  Never-married   
195970    Female           3       1982                   17.0  Never-married   

          childathome  childnotathome  wageincome        weeklyhrscomputer  \
personid                                                                     
195884            NaN             NaN         NaN                      NaN   
195891            NaN             NaN     14000.0  Less than 1 hour a week   
195970            2.0             0.0     52000.0  10 hours or more a week   

                   weeklyhrstv  ...      colenrfeb13      colenroct13  \
personid                        ...                                     
195884                     NaN  ...              NaN              NaN   
195891    3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   
195970    3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   

              colenrfeb14      colenroct14      colenrfeb15      colenroct15  \
personid                                                                       
195884                NaN              NaN              NaN              NaN   
195891    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
195970    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   

              colenrfeb16      colenroct16      colenrfeb17      colenroct17  
personid                                                                      
195884                NaN              NaN              NaN              NaN  
195891    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  
195970    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  

[3 rows x 88 columns]

In [34]:
df.loc[[195884,195891,195970]] # Passing a list of index labels

gender  birthmonth  birthyear  highestgradecompleted  maritalstatus  \
personid                                                                        
195884      Male          12       1981                    NaN            NaN   
195891      Male           9       1980                   12.0  Never-married   
195970    Female           3       1982                   17.0  Never-married   

          childathome  childnotathome  wageincome        weeklyhrscomputer  \
personid                                                                     
195884            NaN             NaN         NaN                      NaN   
195891            NaN             NaN     14000.0  Less than 1 hour a week   
195970            2.0             0.0     52000.0  10 hours or more a week   

                   weeklyhrstv  ...      colenrfeb13      colenroct13  \
personid                        ...                                     
195884                     NaN  ...              NaN              NaN   
195891    3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   
195970    3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   

              colenrfeb14      colenroct14      colenrfeb15      colenroct15  \
personid                                                                       
195884                NaN              NaN              NaN              NaN   
195891    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
195970    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   

              colenrfeb16      colenroct16      colenrfeb17      colenroct17  
personid                                                                      
195884                NaN              NaN              NaN              NaN  
195891    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  
195970    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  

[3 rows x 88 columns]

## Selecting rows using "iloc"

In [37]:
df.iloc[[0,4,5]] # Give a list of row positions- not index labels

gender  birthmonth  birthyear  highestgradecompleted maritalstatus  \
personid                                                                       
100061    Female           5       1980                   13.0       Married   
100583      Male           1       1980                   13.0       Married   
100833      Male           1       1981                   12.0       Married   

          childathome  childnotathome  wageincome        weeklyhrscomputer  \
personid                                                                     
100061            4.0             0.0     12500.0  10 hours or more a week   
100583            4.0             0.0     30000.0  Less than 1 hour a week   
100833            2.0             0.0     39000.0      1 to 3 hours a week   

                    weeklyhrstv  ...      colenrfeb13      colenroct13  \
personid                         ...                                     
100061    11 to 20 hours a week  ...  1. Not enrolled  1. Not enrolled   
100583     3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   
100833     3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   

              colenrfeb14      colenroct14      colenrfeb15      colenroct15  \
personid                                                                       
100061    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
100583    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
100833    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   

              colenrfeb16      colenroct16      colenrfeb17      colenroct17  
personid                                                                      
100061    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  
100583    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  
100833    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  

[3 rows x 88 columns]

In [39]:
df.iloc[0:3] #passing range

gender  birthmonth  birthyear  highestgradecompleted  maritalstatus  \
personid                                                                        
100061    Female           5       1980                   13.0        Married   
100139      Male           9       1983                   12.0        Married   
100284      Male          11       1984                    7.0  Never-married   

          childathome  childnotathome  wageincome        weeklyhrscomputer  \
personid                                                                     
100061            4.0             0.0     12500.0  10 hours or more a week   
100139            2.0             0.0    120000.0      1 to 3 hours a week   
100284            1.0             0.0     58000.0                      NaN   

                    weeklyhrstv  ...      colenrfeb13      colenroct13  \
personid                         ...                                     
100061    11 to 20 hours a week  ...  1. Not enrolled  1. Not enrolled   
100139     3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   
100284    11 to 20 hours a week  ...  1. Not enrolled  1. Not enrolled   

              colenrfeb14      colenroct14      colenrfeb15      colenroct15  \
personid                                                                       
100061    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
100139    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   
100284    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled   

              colenrfeb16      colenroct16      colenrfeb17      colenroct17  
personid                                                                      
100061    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  
100139    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  
100284    1. Not enrolled  1. Not enrolled  1. Not enrolled  1. Not enrolled  

[3 rows x 88 columns]

## Select rows using a condition

In [42]:
df.columns

Index(['gender', 'birthmonth', 'birthyear', 'highestgradecompleted',
       'maritalstatus', 'childathome', 'childnotathome', 'wageincome',
       'weeklyhrscomputer', 'weeklyhrstv', 'nightlyhrssleep', 'satverbal',
       'satmath', 'gpaoverall', 'gpaenglish', 'gpamath', 'gpascience',
       'highestdegree', 'govprovidejobs', 'govpricecontrols', 'govhealthcare',
       'govelderliving', 'govindhelp', 'govunemp', 'govincomediff',
       'govcollegefinance', 'govdecenthousing', 'govprotectenvironment',
       'weeksworked00', 'weeksworked01', 'weeksworked02', 'weeksworked03',
       'weeksworked04', 'weeksworked05', 'weeksworked06', 'weeksworked07',
       'weeksworked08', 'weeksworked09', 'weeksworked10', 'weeksworked11',
       'weeksworked12', 'weeksworked13', 'weeksworked14', 'weeksworked15',
       'weeksworked16', 'weeksworked17', 'colenrfeb97', 'colenroct97',
       'colenrfeb98', 'colenroct98', 'colenrfeb99', 'colenroct99',
       'colenrfeb00', 'colenroct00', 'colenrfeb01', 'col

In [44]:
df.nightlyhrssleep.describe() #statistical info of the nightlyhrssleep column

count    6706.000000
mean        6.637340
std         1.364404
min         0.000000
25%         6.000000
50%         7.000000
75%         8.000000
max        20.000000
Name: nightlyhrssleep, dtype: float64

In [46]:
df.nightlyhrssleep.quantile(0.25)

6.0

In [48]:
df.nightlyhrssleep.quantile(0.05)

4.0

Out of 6706 people responded to this question, 
5 % of people said that they have 4 hrs or less than 4 hrs of sleep

## Selecting the rows who have 4 hrs or less sleep 

In [52]:
low_sleep_people = df.loc[df.nightlyhrssleep<=4]

In [54]:
low_sleep_people.shape

(364, 88)

In [56]:
low_sleep_people.head(2)

gender  birthmonth  birthyear  highestgradecompleted  maritalstatus  \
personid                                                                       
101414     Male           6       1984                   12.0  Never-married   
102454     Male           4       1982                   16.0        Married   

          childathome  childnotathome  wageincome    weeklyhrscomputer  \
personid                                                                 
101414            2.0             0.0     20000.0                  NaN   
102454            NaN             NaN         NaN  1 to 3 hours a week   

                   weeklyhrstv  ...      colenrfeb13      colenroct13  \
personid                        ...                                     
101414    3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   
102454    3 to 10 hours a week  ...  1. Not enrolled  1. Not enrolled   

              colenrfeb14      colenroct14      colenrfeb15  \
personid                                                      
101414    1. Not enrolled  1. Not enrolled  1. Not enrolled   
102454    1. Not enrolled  1. Not enrolled  1. Not enrolled   

                 colenroct15         colenrfeb16      colenroct16  \
personid                                                            
101414       1. Not enrolled     1. Not enrolled  1. Not enrolled   
102454    2. 2-year college   2. 2-year college   1. Not enrolled   

              colenrfeb17         colenroct17  
personid                                       
101414    1. Not enrolled     1. Not enrolled  
102454    1. Not enrolled  2. 2-year college   

[2 rows x 88 columns]

## Select rows based on multiple conditions

In [59]:
low_sleep_people.childathome.describe()

count    293.000000
mean       1.788396
std        1.400685
min        0.000000
25%        1.000000
50%        2.000000
75%        3.000000
max        9.000000
Name: childathome, dtype: float64

In [63]:
low_sleep_3_plus_children = df.loc[(df.nightlyhrssleep <= 4) & (df.childathome >= 3)]

In [65]:
low_sleep_3_plus_children.shape

(82, 88)

## Select rows - 3 conditions

In [68]:
low_sleep_3_plus_children_females = df.loc[(df.nightlyhrssleep <=4) & \
                                            (df.childathome >=3) & \
                                            (df.gender == "Female")
                                           ]

In [70]:
low_sleep_3_plus_children_females.shape

(57, 88)

## Selecting rows and columns using "loc"

In [73]:
low_sleep_3_plus_child = df.loc[
                                (df.nightlyhrssleep <=4) & \
                                (df.childathome >=3),\
                                ['nightlyhrssleep', 'childathome']
                                ]

In [77]:
low_sleep_3_plus_child.head()

nightlyhrssleep  childathome
personid                              
119754                4.0          4.0
141531                4.0          5.0
152706                4.0          4.0
156823                1.0          3.0
158355                4.0          4.0